### PR TITLE
Cythonise IVOM & K-mer hashing

### PR DESCRIPTION
**Pull request by kdm9**
This is work in progress, don't merge yet.

I've rewritten the hashing to use kmer count vectors, and re-done the IVOM function  all in cython. Does about 2-5 3kb windows every second, including hashing. How does that compare time-wise?

### CODE REVIEW

File: frisk/kmerhash.pyx [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    return kmer_freqs
+
+def ivom(dict kmer_freqs not None, int alphabet_sz=4):
+    cdef int min_k = min(kmer_freqs.keys())
+    cdef int max_k = max(kmer_freqs.keys())
+    cdef unsigned long long subkmer = 0
+    cdef float count = 0
+
+    maxmer_ivoms = np.zeros(kmer_freqs[max_k].shape, dtype=np.float)
+    sums = {k: v.sum() for k, v in kmer_freqs.items()}
```
Comment (kdmurray91) [inline frisk/kmerhash.pyx]: These are freqs so this will be 1. that's a bug

**Comment by Adamtaranto** [comment]
I'll add a timer for window IVOM & KLI processing. I'd guess about 1-2 5000bp windows per second.

computeKmers() when run in 'genomeMode' is also a time issue. Not sure if there is a way around slicing up the genome one kmer at a time to do initial kmer census.

**Comment by kdmurray91** [comment]
If you want to test it out, I'm using https://gist.github.com/19e3df51c6e4f3af642a

Also, you need to do `python setup.py build_ext --inplace` before running the above.

**Comment by kdmurray91** [comment]
Also this hasn't been optimised at all, just compiled to C w/ cython. Which isn't magic, I was just trying to re-implement IVOM such that I understood it. We first check that the window IVOM you get for two sequences is the same (which it won't be yet as there are a couple of bugs).

File: frisk/kmerhash.pyx [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        weights /= weights.cumsum()
+
+        last_ivom = 0.0
+        for x in range(min_k - 1, max_k):
+            last_ivom = weights[x] * probs[x] + (1 - weights[x] * last_ivom)
+        maxmer_ivoms[kmer] = last_ivom
+    return maxmer_ivoms / maxmer_ivoms.sum()
+
+
+def kli(cnp.ndarray gen_ivom, cnp.ndarray window_ivom):
```
Comment (Adamtaranto) [inline frisk/kmerhash.pyx]: KLI should be the sum of (w * np.log2(w/g)) as calculated for each unique k-mer in the window, where w = weighted probability (IVOM) of a kmer in window, and g = IVOM of the same kmer in genome.

Currently, you're summing up all the w-IVOMs and g-IVOMs across all kmers in a window and calculating KLI in one go.

File: setup.py [hunk 3]
```diff
@@ -55,6 +56,7 @@
     version=versioneer.get_version(),
     install_requires=install_requires,
     tests_require=test_requires,
+    ext_modules=cythonize('frisk/kmerhash.pyx'),
```
Comment (Adamtaranto) [inline setup.py]: I had to add "include_dirs=[numpy.get_include()]," to get this to work.

File: frisk/kmerhash.pyx (hunk 2 above)
Comment (kdmurray91) [inline frisk/kmerhash.pyx]: Yeah, as far as I can tell that is actually what is happening. numpy just vectorises the operations.

So, `w/g` is a vector of length `len(w)`, with each element equal to `w_i/g_i`. Likewise with np.log2, that gives you a vector, and `w*np.log`. Then we sum over that vector skipping NaN (which is what nansum does).

File: setup.py (hunk 3 above)
Comment (kdmurray91) [inline setup.py]: Ah, yes, forgot that.

File: frisk/kmerhash.pyx (hunk 1 above)
Comment (kdmurray91) [inline frisk/kmerhash.pyx]: Actually, this doesn't matter. If you pass the frequency or the non-normalised version it gives the same answer. The normalisation just happens internally.

**Pull request closed without merging**